dlgr/griduniverse/maze.py

```diff
@@ -0,0 +1,119 @@
+            if yy == y:
+                ver[y][max(x, xx)] = "  "
+            stack.append((xx, yy))
+            visited[yy][xx] = 1
+
+    # Convert the maze to a list of wall cell positions.
+    the_rows = ([j for i in zip(hor, ver) for j in i])
+    the_rows = [list("".join(j)) for j in the_rows]
+    maze = [item == '*' for sublist in the_rows for item in sublist]
+    positions = []
+    for idx, value in enumerate(maze):
+        if value:
+            positions.append([idx / columns, idx % columns])
+
+    return positions
+
+
+def prune(walls, density, contiguity):
+    """Prune walls to a labyrinth with the given density and contiguity."""
+    num_to_prune = int(round(len(walls) * (1 - density)))
+    num_pruned = 0
+    count = 0
+    while num_pruned < num_to_prune:
+        count += 1
+        to_prune = _classify_terminals(
+            walls, limit=num_to_prune - num_pruned
+        )
+        walls = [w for i, w in enumerate(walls) if i not in to_prune]
+        if len(to_prune) == 0:
+            break
+        num_pruned += len(to_prune)
+
+    num_to_prune = int(round(len(walls) * (1 - contiguity)))
+    to_prune = set(random.sample(range(len(walls)), num_to_prune))
+    walls = [w for i, w in enumerate(walls) if i not in to_prune]
+
+    return walls
+
+
+def _classify_terminals(walls, limit=None):
+    found = [set()]
+    unmatched = {}
+    enumerated_walls = tuple(enumerate(walls))
+    position_map = {tuple(w.position): i for i, w in enumerated_walls}
+
+    for i, w in enumerated_walls:
+        neighbors = set()
+        for adj in ([1, 0], [-1, 0], [0, 1], [0, -1]):
+            neighbor = tuple(p1 + p2 for p1, p2 in zip(w.position, adj))
+            if neighbor in position_map:
+                neighbors.add(neighbor)
+        if len(neighbors) <= 1:
+            found[0].add(i)
+            if limit and len(found[0]) >= limit:
+                break
+        elif len(neighbors) == 2:
+            n_indexes = {position_map[i] for i in neighbors}
+            j = 0
+            while j < len(found):
+                if n_indexes.intersection(found[j]):
+                    if len(found) < j + 2:
+                        found.append(set())
+                    up_next = found[j + 1]
+                    up_next.add(i)
+                    break
+                j += 1
+            else:
+                unmatched[i] = n_indexes
+
+    for index in unmatched:
+        j = 0
+        n_indexes = unmatched[index]
+        while j < len(found):
+            if n_indexes.intersection(found[j]):
+                if len(found) < j + 2:
+                    found.append(set())
+                up_next = found[j + 1]
+                up_next.add(index)
+                break
+            j += 1
+
+    to_prune = set()
+    count = 0
+    for entries in found:
+        count += 1
+        if len(to_prune) < limit:
+            to_prune = to_prune.union(set(itertools.islice(entries, limit - len(to_prune))))
```

Comment:
I think the issue where a sparse maze gets more dense toward the bottom may be the result of slicing the results for the last set of walls. A set of wall positions shouldn't necessarily be ordered from top to bottom, but it wouldn't surprise me if it was (the prior implementation had the same issue, but didn't use sets). As a result, it would probably be better to take a random sample instead of a slice when the number needed is larger than the set.